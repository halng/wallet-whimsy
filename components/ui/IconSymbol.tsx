/*
 * Copyright 2025 Hal Ng
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import MaterialIcons from '@expo/vector-icons/MaterialIcons';
import { SymbolWeight } from 'expo-symbols';
import React from 'react';
import { OpaqueColorValue, StyleProp, ViewStyle } from 'react-native';

// Add your SFSymbol to MaterialIcons mappings here.
const MAPPING = {
  // See MaterialIcons here: https://icons.expo.fyi
  // See SF Symbols in the SF Symbols app on Mac.
  'house.fill': 'home',
  'paperplane.fill': 'send',
  'chevron.left.forwardslash.chevron.right': 'code',
  'chevron.right': 'chevron-right',
} as Partial<
  Record<
    import('expo-symbols').SymbolViewProps['name'],
    React.ComponentProps<typeof MaterialIcons>['name']
  >
>;

export type IconSymbolName = keyof typeof MAPPING;

/**
 * An icon component that uses native SFSymbols on iOS, and MaterialIcons on Android and web. This ensures a consistent look across platforms, and optimal resource usage.
 *
 * Icon `name`s are based on SFSymbols and require manual mapping to MaterialIcons.
 */
export function IconSymbol({
  name,
  size = 24,
  color,
  style,
}: {
  name: IconSymbolName;
  size?: number;
  color: string | OpaqueColorValue;
  style?: StyleProp<ViewStyle>;
  weight?: SymbolWeight;
}) {
  return <MaterialIcons color={color} size={size} name={MAPPING[name]} style={style} />;
}
