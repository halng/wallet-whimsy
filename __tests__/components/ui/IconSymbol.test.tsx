/*
 * Copyright 2025 Hal Ng
 * Licensed under the Apache License, Version 2.0 (the License);
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an AS IS BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { render } from '@testing-library/react-native';

import { IconSymbol } from '@/components/ui/IconSymbol';

describe('IconSymbol', () => {
  it('matches snapshot with default props', () => {
    const { toJSON } = render(<IconSymbol name="house.fill" color="black" />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('matches snapshot with custom size and style', () => {
    const { toJSON } = render(
      <IconSymbol name="paperplane.fill" color="red" size={32} style={{ marginRight: 10 }} />,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('matches snapshot for different icon mappings', () => {
    const { toJSON } = render(
      <>
        <IconSymbol name="chevron.right" color="blue" />
        <IconSymbol name="money.bill.transfer" color="green" />
        <IconSymbol name="account.balance" color="purple" />
      </>,
    );
    expect(toJSON()).toMatchSnapshot();
  });
});
