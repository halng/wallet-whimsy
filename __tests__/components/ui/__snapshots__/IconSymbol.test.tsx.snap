// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IconSymbol matches snapshot for different icon mappings 1`] = `
[
  <ViewManagerAdapter_SymbolModule
    animated={false}
    colors={[]}
    name="chevron.right"
    resizeMode="scaleAspectFit"
    style={
      [
        {
          "height": 24,
          "width": 24,
        },
        [
          {
            "height": 24,
            "width": 24,
          },
          undefined,
        ],
      ]
    }
    tintColor="blue"
    type="monochrome"
    weight="regular"
  />,
  <ViewManagerAdapter_SymbolModule
    animated={false}
    colors={[]}
    name="money.bill.transfer"
    resizeMode="scaleAspectFit"
    style={
      [
        {
          "height": 24,
          "width": 24,
        },
        [
          {
            "height": 24,
            "width": 24,
          },
          undefined,
        ],
      ]
    }
    tintColor="green"
    type="monochrome"
    weight="regular"
  />,
  <ViewManagerAdapter_SymbolModule
    animated={false}
    colors={[]}
    name="account.balance"
    resizeMode="scaleAspectFit"
    style={
      [
        {
          "height": 24,
          "width": 24,
        },
        [
          {
            "height": 24,
            "width": 24,
          },
          undefined,
        ],
      ]
    }
    tintColor="purple"
    type="monochrome"
    weight="regular"
  />,
]
`;

exports[`IconSymbol matches snapshot with custom size and style 1`] = `
<ViewManagerAdapter_SymbolModule
  animated={false}
  colors={[]}
  name="paperplane.fill"
  resizeMode="scaleAspectFit"
  style={
    [
      {
        "height": 24,
        "width": 24,
      },
      [
        {
          "height": 32,
          "width": 32,
        },
        {
          "marginRight": 10,
        },
      ],
    ]
  }
  tintColor="red"
  type="monochrome"
  weight="regular"
/>
`;

exports[`IconSymbol matches snapshot with default props 1`] = `
<ViewManagerAdapter_SymbolModule
  animated={false}
  colors={[]}
  name="house.fill"
  resizeMode="scaleAspectFit"
  style={
    [
      {
        "height": 24,
        "width": 24,
      },
      [
        {
          "height": 24,
          "width": 24,
        },
        undefined,
      ],
    ]
  }
  tintColor="black"
  type="monochrome"
  weight="regular"
/>
`;
